feat: cattach

huerni opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Task-specific I/O metadata enabling per-task input routing and per-task output delivery within multi-task jobs
  * Step metadata query to retrieve step and node assignment details
* **Improvements**
  * More resilient stream handling with reconnect support and guarded reply delivery
  * Output queue memory management with a hard byte cap and overflow-dropping behavior
  * Preserves queued output across reconnects and bounds reconnect attempts
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

huerni committed:
todo

huerni committed:
feat: add m_cfored_stream_proxy_map_

huerni committed:
feat: io forward when cfored restart

huerni committed:
feat: cattach recive data

huerni committed:
feat: output data size

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -655,7 +670,22 @@ void CforedClient::TaskEnd(pid_t pid) {
 void CforedClient::TaskOutPutForward(const std::string& msg) {
   CRANE_TRACE("Receive TaskOutputForward len: {}.", msg.size());
+
+  if (msg.size() > kMaxOutputQueueBytes) {
```
这里截断下？永远取最新的输出吧

L-Xiafeng reviewed (commented):
这里如果一直连不上，这个任务就卡在重连了吧？

huerni commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -30,6 +30,15 @@ using crane::grpc::StreamTaskIORequest;
+      [this](const uvw::async_event&, uvw::async_handle&) {
+        while (m_wait_reconn_ && !m_stopped_) {
+          InitChannelAndStub(m_cfored_name_);
+          std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
多个task的任务确实没考虑  我rebase后整体看看

huerni committed:
refactor

huerni committed:
merge master

huerni committed:
feat: maxReconnect

huerni committed:
refactor cforedclient reconnect

huerni committed:
fix: Fix reconnection failure caused by leftover events from the previous stream in m_cq_

huerni committed:
feat: cattach connect step

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
refactor: rename TaskScheduler refs to JobScheduler and remove obsolete task methods

- Rename `TaskScheduler::QueryStepAndNodeRegex` to `JobScheduler::QueryStepAndNodeRegex` and update internal map references from `m_running_task_map_` to `m_running_job_map_`
- Replace hardcoded step ID `1` with `kPrimaryStepId` constant for clarity
- Remove obsolete `TaskScheduler::QueryTaskUseId` method and `JOB_META_REQUEST` handler that duplicated existing functionality
- Update `CtldGrpcServer` to use `g_job_scheduler` instead of `g_task_scheduler` for step queries
- Replace `WriteTaskMetaReply` with `WriteStepMetaReply` to better reflect step-level semantics
- Fix log message in `CforedStream` to correctly say `JOB_REQUEST` and use `job->JobId()` instead of `task->TaskId()`

huerni committed:
feat(Craned): add cfored reconnection logic in AsyncSendRecvThread

Refactor `AsyncSendRecvThread_` to support automatic reconnection to
cfored when the connection is lost. The state machine loop is now
wrapped in an outer retry loop that checks `m_wait_reconn_` and
`m_reconnect_attempts_` before re-establishing the gRPC stream.

Key changes:
- Add outer reconnect loop with exponential backoff up to
  `kMaxReconnectIntervalSec` seconds between attempts
- Enforce a `kMaxReconnectAttempts` limit; on exhaustion, mark all
  forwarded tasks as stopped, terminate the step via
  `TerminateStepAsync`, and exit the thread
- Move per-session variables (stream, context, write_pending, etc.)
  inside the reconnect loop so they are re-initialized on each attempt
- On stream failure, transition to a reconnecting state instead of
  immediately terminating, allowing the outer loop to retry

huerni committed:
fix(Craned): improve CforedClient reconnect and output drain logic

- Track dequeued bytes for TASK_OUTPUT to correctly maintain
  `m_output_queue_bytes_` counter when items are consumed
- Replace `goto exited` with early `return` on reconnect to preserve
  queued output data; only set `m_output_drained_` on normal exit
- Add pre-write reconnect check to avoid issuing writes after a
  reconnect signal is detected mid-wait
- Implement exponential backoff with an upper bound of
  `kMaxReconnectIntervalSec` before each reconnect attempt
- Check gRPC channel state after backoff and skip reconnect if the
  channel is in TRANSIENT_FAILURE or SHUTDOWN; proceed when IDLE,
  CONNECTING, or READY
- Improve log messages to distinguish normal vs. reconnect exits and
  report attempt counts and wait intervals clearly
- Fix minor typo: "Markd" → "Marked" in error log

huerni committed:
fix(Craned): set output_drained flag on all thread exit paths

Ensure `m_output_drained_` is stored as `true` in both reconnect exit
branches of `CleanOutputQueueAndWriteToStreamThread_`, not just on
normal exit. Previously, early returns during reconnect skipped setting
the flag, which could leave waiters blocked indefinitely.

huerni committed:
fix(CforedClient): delay reconnect sleep until channel state is checked

Move `sleep_for` after the channel state check so reconnection is
only delayed when the channel is not ready, avoiding unnecessary
waits when the channel recovers quickly. Also simplify the state
condition from checking specific failure states to `!= GRPC_CHANNEL_READY`
for broader coverage.

huerni committed:
feat(proto): add task_id to StreamStepIO for targeted I/O routing

Add `task_id` field to `TaskOutputReq` to identify which task produced
the output, and an optional `task_id` to `TaskInputReq` to allow stdin
delivery to a specific task instead of broadcasting to all tasks.

Update `CforedClient` to capture and pass `task_id` when forwarding
task stdout via both pipe and TTY handles, aligning the implementation
with the updated proto contract.

Also normalize trailing comment alignment in Crane.proto by removing
extra spaces used for column-alignment.

huerni committed:
fix: reduce kMaxOutputQueueBytes from 1MB to 1KB

Decrease the maximum output queue size constant from 1 * 1024 * 1024
(1MB) to 1024 bytes (1KB) to limit memory usage for task output
buffering. Also add a missing blank line in CforedClient.cpp for
readability.

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
fix(JobScheduler): add null pointer checks in QueryStepAndNodeRegex

Add null checks for `job` and `primary_step` pointers in
`QueryStepAndNodeRegex` to prevent potential null pointer dereferences.
Log debug messages when either pointer is null and return false early
to handle these edge cases gracefully.

huerni committed:
fix(Supervisor): move set_task_id call into correct stdout branch scope

The `payload->set_task_id(req.task_id)` call was placed after the
if-else block, outside the scope where `payload` was defined. This
caused `payload` to be inaccessible at that point. Move the call
inside the `if (req.is_stdout)` branch where the `payload` variable
is properly in scope.

huerni committed:
fix: increase max output queue size from 1KB to 10MB

The `kMaxOutputQueueBytes` constant was set to only 1024 bytes (1 KB),
which is likely too small for real-world job output buffering. Increase
it to 10 MB to prevent premature queue overflow and data loss when
handling larger output streams.

github-actions[bot] committed:
style: auto format with clang-format.

L-Xiafeng commented on src/CraneCtld/RpcService/CtldGrpcServer.cpp:
```diff
@@ -441,8 +484,38 @@ grpc::Status CtldForInternalServiceImpl::CforedStream(
                 "Failed to send msg to cfored {}. Connection is broken. "
                 "Exiting...",
```
这种分支的日志都改下?

L-Xiafeng commented on protos/Crane.proto:
taskerr也需要支持

huerni replied:
认同这个需求，`TaskErrOutputReq` 也应该携带 `task_id` 以支持 stderr 的 per-task 过滤。由于涉及 proto 变更和前后端同步修改，计划在后续 PR 中处理。

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -522,6 +523,14 @@ void CforedClient::CleanOutputQueueAndWriteToStreamThread_(
+        tracked_len = std::get<IOFwdRequest>(fwd_req.data).len;
+        m_output_queue_bytes_.fetch_sub(tracked_len,
+                                        std::memory_order::relaxed);
+      }
```
为什么写在这里？不应该写在下面的std::visit里面？

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -585,257 +612,320 @@ void CforedClient::AsyncSendRecvThread_() {
+          bool has_x11 = false;
+          {
+            absl::MutexLock lock(&m_mtx_);
+            for (auto& [id, _] : m_x11_fd_info_map_) {
```
std::views:values

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -585,257 +612,320 @@ void CforedClient::AsyncSendRecvThread_() {
+      // Channel is IDLE, CONNECTING, or READY: proceed to create new stream
+      CRANE_INFO("Channel state {} usable, creating new stream...",
+                 static_cast<int>(ch_state));
+      m_wait_reconn_ = false;
```
这一块的逻辑不太对，就算第一次启动也会输出reconn的msg，而且success后应该重置reconn attempt？

huerni replied:
第一次启动m_wait_reconn_是false，不会进入reconn逻辑吧？

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -559,8 +569,25 @@ void CforedClient::CleanOutputQueueAndWriteToStreamThread_(
+
+      // One more check before issuing the write
+      if (m_wait_reconn_.load(std::memory_order::acquire)) {
+        CRANE_TRACE("CleanOutputQueueThread: reconnect exit");
```
什么叫reconnect exit？应该叫 exit on disconnection

L-Xiafeng commented on src/CraneCtld/RpcService/CtldGrpcServer.h:
```diff
@@ -180,6 +197,29 @@ class CforedStreamWriter {
+
+ private:
+  std::mutex mtx_;
+  std::shared_ptr<CforedStreamWriter> writer_;
```
没有必要用mutex？一个cfored+step对应一个map entry，对于单个cfored，是单线程的，应该不会有datarace？

huerni replied:
`StreamWriterProxy` 的 `SetWriter` 在 cfored 重连时由新的流处理线程调用，而 `WithWriter` 可能被 JobScheduler 线程池中的回调（如 `cb_step_res_allocated`）调用，两者在不同线程，mutex 是必要的。

huerni committed:
fix(JobScheduler): correct nodelist assignment in QueryStepAndNodeRegex method
fix(CtldGrpcServer): improve error handling and logging for stream disconnection

huerni committed:
fix(CforedClient): track bytes for TASK_OUTPUT only on stdout and improve logging on disconnection

huerni committed:
fix(CforedClient): update termination cause for cfored connection failure

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
fix: reorder fields in StepMetaReq and improve comments in related methods

huerni committed:
feat: add cattach command documentation in English and Chinese

huerni committed:
fix(CtldGrpcServer): update error message for cfored connection issues to indicate waiting for reconnection

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -513,7 +514,7 @@ void CforedClient::CleanOutputQueueAndWriteToStreamThread_(
 
   // Make sure before exit all output has been drained.
-  while (!m_stopped_ || ok) {
+  while (!m_wait_reconn_ && (!m_stopped_ || ok)) {
```
m_wait_reconn_.load(std::memory_order::acquire)

L-Xiafeng commented:
LGTM

huerni committed:
fix(CforedClient): improve reconnect logic and enhance comments for clarity

huerni committed:
merge master

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -585,257 +613,320 @@ void CforedClient::AsyncSendRecvThread_() {
-        state = State::Unregistering;
+      // Exponential backoff with upper bound of kMaxReconnectIntervalSec
+      int interval =
+          static_cast<int>(std::min(attempts, kMaxReconnectIntervalSec));
```
并不是Exponential backoff

L-Xiafeng commented on src/Craned/Supervisor/CforedClient.cpp:
```diff
@@ -585,257 +613,320 @@ void CforedClient::AsyncSendRecvThread_() {
-      continue;
+      // Channel is IDLE, CONNECTING, or READY: proceed to create new stream
+      CRANE_INFO("Channel state {} usable, creating new stream...",
+                 static_cast<int>(ch_state));
```
不重置m_reconnect_attempts_？

L-Xiafeng commented on src/CraneCtld/JobScheduler.cpp:
```diff
@@ -5015,6 +5015,39 @@ void JobScheduler::QueryJobsInRam(
+  auto iter = m_running_job_map_.find(job_id);
+  if (iter == m_running_job_map_.end()) return false;
+
+  if (step_id == 0) return false;
```
这个应该放到前面？

L-Xiafeng commented on protos/Crane.proto:
```diff
@@ -878,6 +888,13 @@ message StreamCtldReply {
+  message StepMetaReply {
+    bool ok = 1;
+    string failure_reason = 2;
+    StepToCtld step = 3;
```
为什么要传一个完整的StepToCtld ？

huerni committed:
feat: enhance StepToCtld with craned_task_map and update related methods for task management

huerni committed:
feat: add CattachStepInfo message and update StepMetaReply to include step_info

huerni committed:
feat: add NodeTasks message and craned_task_map to CattachStepInfo for enhanced task management

huerni committed:
refactor: remove craned_task_map from StepMetaReply and update CattachStepInfo for task routing

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
feat: add validation for interactive job types in CforedStream method

huerni committed:
feat: add validation for cfored_name in CforedStream method

huerni commented:
@L-Xiafeng 线下已跑过，可approval 等待合入
